[DRAFT] extraneous components and version range constraints

jkowalleck opened the pull request:
Sketch/proposal for  #321

implementing with `components`, because the objects referenced/required are actually used at runtime and therefore are considered a "component". 



- [x] sketch JSON schema  
  - properties and assert
  - test cases 
- [x] sketch XML schema
  - properties. assert would require XSD1.1 which is not broadly implemented, yet.
  - test cases
- [ ] sketch ProtoBuff schema  
  SKIPPED THIS FOR NOW as I do not have a lot of practice with PB schema
  Please help :-)

stevespringett commented:
Thank you for putting this together. Few questions.

How is `extraneous` different from a `scope` with a value of `excluded`?

The versionRange is interesting. I like it, but it will not work in all situations. For example, if a extraneous component is Windows and the range is 2000, XP, Vista, 7, 8, 10, and 11. I think we'll need another way to specify possible versions.

stevespringett commented:
> I could add examples for popular operating systems and runtime-engines to bom:versionRangeType, if you want.

I don't think that's necessary. I'm just wondering how many people would understand that marketing names have actual version numbers.

prabhu commented on schema/bom-1.6.schema.json:
```diff
@@ -913,6 +919,12 @@
           "description": "Specifies the scope of the component. If scope is not specified, 'required' scope SHOULD be assumed by the consumer of the BOM.",
           "default": "required"
         },
+        "isExtraneous": {
```
If the environment provides a component, then it is best to capture it under formulation instead of having such boolean flags.

jkowalleck replied:
why should it be under "formulation"? 
form,ulation is build-time, while this domain here may also be runtime.

prabhu replied:
Wouldn't such runtime-only components not captured in an OBOM?

jkowalleck replied:
how would you make it clear in an obom, that the build result descibed in an SBOM needs windows 11 patch 23H2 on runtime? Or that it  requires python 3.11 or later as an interpreter? 
I dont know ... this is why this PR exists, to discuss these options and maube provide a solution :-) 

read more here in the sectiopn "usecases" of https://github.com/CycloneDX/specification/issues/321

prabhu replied:
Thank you. I should've read the use cases first, which helped. I would have used two separate documents and linked them via external references. However, I see the appeal of having a single combined document for both SBOM and its operating environment. Perhaps component.type = "platform" indicates runtime without the need for the additional boolean flag?

jkowalleck commented:
Dropped this issue/request from the 1.6 milestone goals, and moved it to 1.7 for the following reasons: did not finish in time for 1.6, needs further discussion. 


need to migrate this PR after 1.6 was finished/published/merged ...

jkowalleck commented on tools/src/test/resources/1.6/informal-invalid-component-versionRange-non-extraneous-explicit.xml:
```diff
@@ -0,0 +1,16 @@
+    <components>
+        <component type="library" isExtraneous="false">
+            <name>InvalidVersions</name>
+            <versionRange><![CDATA[>=9.0.0|<10.0.0]]></versionRange>
```
I just took the version-part of VERS.
maybe this was wrong, need to read the specs of verts again ...

jkowalleck commented:
superseded by #586
